platform/view/services/storage/vault/vaultstore_test.go

```diff
@@ -84,57 +163,90 @@ func NewOffsetPagination(offset int, pageSize int) driver.Pagination {
 	return offsetPagination
 }
 
-func getAllTxStatuses(store driver.VaultStore, pagination driver.Pagination) ([]driver.TxStatus, error) {
-	p, err := store.GetAllTxStatuses(context.Background(), pagination)
-	if err != nil {
-		return nil, err
-	}
-	statuses, err := collections.ReadAll(p.Items)
+func NewKeysetPagination(offset int, pageSize int, sqlIdName common2.FieldName, idFieldName pagination.PropertyName[string]) driver.Pagination {
+	keysetPagination, err := pagination.KeysetWithField[string](offset, pageSize, sqlIdName, idFieldName)
 	if err != nil {
-		return nil, err
+		Expect(err).ToNot(HaveOccurred())
 	}
-	return statuses, nil
+	return keysetPagination
 }
 
 func testPagination(store driver.VaultStore) {
-	RegisterFailHandler(Fail)
 	err := store.SetStatuses(context.Background(), driver.TxStatusCode(valid), "",
 		"txid1", "txid2", "txid10", "txid12", "txid21", "txid100", "txid200", "txid1025")
 	Expect(err).ToNot(HaveOccurred())
 
+	pi := pagination.NewDefaultInterpreter()
+
 	for _, item := range matrix {
-		for page := 0; page < len(item.matcher); page++ {
-			statuses, err := getAllTxStatuses(store, item.pagination)
+		if len(item.sqlBackward) == 0 {
+			item.sqlBackward = item.sqlForward
+			item.argsBackward = item.argsForward
+		}
+		pag := item.pagination
+		page := 0
+		for ; true; page++ {
+			// We don't need to build the query here but we do it just to test that it was build correctly
+			query, args := q.Select().
+				AllFields().
+				From(q.Table("test")).
+				Paginated(pag).
+				FormatPaginated(nil, pi)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(query).To(Equal(item.sqlForward[page]))
+			Expect(args).To(ConsistOf(item.argsForward[page]))
+
+			p, err := store.GetAllTxStatuses(context.Background(), pag)
+			Expect(err).ToNot(HaveOccurred())
+			statuses, err := collections.ReadAll(p.Items)
 			Expect(err).ToNot(HaveOccurred())
+			// Test we get 0 statuses when we reach the end
 			if len(statuses) == 0 {
 				break
 			}
+			Expect(page).To(BeNumerically("<", len(item.matcher)))
+			Expect(err).ToNot(HaveOccurred())
 			Expect(statuses).To(item.matcher[page])
-			item.pagination, err = item.pagination.Next()
+
+			pag, err = p.Pagination.Next()
 			Expect(err).ToNot(HaveOccurred())
 		}
-		for page := len(item.matcher) - 1; 0 <= page; page-- {
-			item.pagination, err = item.pagination.Prev()
+		// test that we read everything
+		Expect(page).To(BeNumerically("==", len(item.matcher)))
```

Comment:
keep a separate counter like `totalPages` or something like that. It's fine to have a second variable. It will make the code more readable to have a trivial for loop and one extra counter.